### PR TITLE
Basic integration testing framework

### PR DESCRIPTION
**Pull request by dead10ck**
This PR sets up a very (very!) basic set up for integration testing this crate. It takes advantage of Docker, `docker-compose`, and Travis to run integration tests on the parts of this library that actually need a running Kafka server to execute. The details are described in #138.

There are still outstanding issues with the approach:

* This set up uses a single-broker Kafka server—to really be useful, the tests should run on a multi-broker setup, as this is the more realistic environment that production Kafka instances are set up in. I have not yet been able to figure out how to get the `wurstmeister/kafka` Docker image to work with a multi-broker setup, however.
    * **Edit:** I have a working setup with 3 brokers! Unfortunately, it required a patch to the `wurstmeister/kafka` Docker image to get working properly for Kafka 0.8. I've [submitted a PR](https://github.com/wurstmeister/kafka-docker/pull/160) that will fix it, but until it's merged and backported to the 0.8 image, my multi-broker setup will live in [this branch](https://github.com/dead10ck/kafka-rust/tree/multi-broker-test). I'll merge it into this branch when the patch is backported.
* Since this crate currently does not have any administrative functionality, like creating topics, it relies on creating a fixed topic on the broker before executing the tests, which may lead to confusion and potentially the Kafka servers getting into weird states while the tests run. To mitigate this, I've made tests run serially, rather than in parallel, but this should be improved, if possible.

However, these problems notwithstanding, I think it is still useful to have these very basic tests and the framework in place so that the most crucial functionality (querying the Kafka server for metadata, sending messages through a producer, and receiving messages from a consumer) is tested at least at a basic level, and so others can begin filling in holes as they work.

Once this PR is merged, the flow will change a bit.

### Local work
When working locally, you must have Docker and `docker-compose` installed. There is a script in the root directory named `test`, which you will run when you want to run tests. It will automatically set up the Docker environment for Kafka, and then run `cargo test` for each Kafka version specified. For example, running `./test 0.8.2.2 0.9.0.0` will run the tests against Kafka versions 0.8.2.2 and 0.9.0.0. The arguments should correspond to a Docker tag of the `wurstmeister/kafka` image, so any valid tag will do—i.e., `./test latest` will run against the most recent Kafka version published in the `wurstmeister/kafka` repo. If there are no versions specified, it runs the tests on a set of default versions.

### Remotely
When Travis runs, it will take advantage of its matrix functionality to start up a job to run the tests in each version specified in the `.travis.yml`, currently set to the most current Kafka versions in the last 3 major releases (0.8, 0.9, and 0.10).

## Other changes
There are a couple of other minor changes:

1. Formatted the code with `cargo fmt`. Sorry about the noise here. Reviewing will be easier if it's done commit-by-commit
2. Replaces `SipHasher` with `DefaultHasher`, as the former has been deprecated

### CODE REVIEW

File: examples/topic-metadata.rs [hunk 1]
```diff
@@ -36,7 +36,10 @@ struct Offsets {
```
Comment (xitep) [inline examples/topic-metadata.rs]: after updating my rust set up, I realized this small program gives the following compilation warning:

```
warning: unused `#[macro_use]` import, #[warn(unused_imports)] on by default
 --> examples/topic-metadata.rs:4:1
  |
4 | #[macro_use]
  | ^^^^^^^^^^^^
```

may i ask you to remove the unused macro item here as part of this PR?

File: .travis.yml [hunk 2]
```diff
@@ -17,10 +27,18 @@ script:
...
+
 after_success:
   - travis-cargo --only stable doc-upload
 
 env:
   global:
+    - RUST_TEST_THREADS: 1
     - TRAVIS_CARGO_NIGHTLY_FEATURE="nightly"
     - secure: zK1SJqtVc2Ha/onG/QvjLfz+JHqxfhLmOJuKpzfi/CiZwCUdVqeJdOd48jBON5vnrJ4LGKiEnydgaEtLYFYUb7Xkvxa3Zk5SeB+EmCVzobUSSjSodtx/HlIWP6HRdd8RFnZR8e945BvXt0bOghzJYNpDMWZxn7gTjm2ZM3vaJUwbZqO+9nAWoERnxS/6AgVuuALgcq+FgmOVTG/hXSWQZvmyNzO/6upIOJ3nBHHtBtfRhTrszk+/fEzmYyNWWXGyD65SaQTvJ4Wt8nnykpeGVYXo+SEtpypmuOxnnNTVXC+6BAK6Q3i1Z65NNKDNyjYEZMMWHIiTDLXuLiHp9Bw9hs5sR6DhXNS84zdYmssylr8TlNUXXhU8IR1+c7R5++HbcopAYFkEZonF7lcL9T9BwbZqFk6wcQKSJrw7N3btfbLtgULfjvBbWrA0md4iqce7x0V3h6zo6BBbhlbxxa3FeAUZ9kMopq0SIjSRHHW/NbAfT9b8tthlRTgqx/eFbdil6pHjtSCdDTN8+V+iaA7WqPycscRo3JhHXunlqJTesA71bw2ovLBjJ6OheBSnYAZ+gYLKF56P3QrJZ6uAIt6kZkKrORoVygy6HRCZDTiNf+5NkDw4AgVULeGfRwqoLnhQOjW3yChbG/Ur4V9Ascr9RVlwRj/6tpX6A5d3UFVn3WA=
+  matrix:
```
Comment (xitep) [inline .travis.yml]: can we have a blank line before `matrix` here?

**Comment by dead10ck** [comment]
I set up a separate PR for the `cargo fmt`; I'll work on the others when I get some time. Thanks for the review!

**Comment by dead10ck** [comment]
@spicavigo Ok, I've made all the changes requested. Some points:

* The integration tests now require the `integration_tests` feature enabled to compile and run. The `run-all-tests` script enables this feature.
* For now, I've decided to leave `SipHasher` as-is with the warnings, and I'll file a ticket to look into alternatives.
* I've moved the integration testing-related files/scripts into the `tests` directory; doing this required some hacky changes to the `run-all-tests` script, since both `docker-compose` and `cargo` want to be at the root directory, but it works.

File: src/client/state.rs [hunk 3]
```diff
@@ -474,7 +474,8 @@ mod tests {
                            assert_eq!(broker.map(|b| b.1), state.find_broker(topic, id));
                            (id, broker)
                        })
-                       .collect::<Vec<_>>()[..]);
+                       .collect::<Vec<_>>()
+                        [..]);
     }
```
Comment (xitep) [inline src/client/state.rs]: `cargo fmt` seems to suggest the original was just fine. maybe just unintended newline here?

File: tests/integration/consumer_producer/mod.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        .with_offset_storage(kafka::consumer::GroupOffsetStorage::Kafka)
+}
+
+/// Return a ready Kafka consumer with all default settings
+pub fn test_consumer() -> Consumer {
+    test_consumer_builder().create().unwrap()
+}
+
+mod producer;
+mod consumer;
```
Comment (xitep) [inline tests/integration/consumer_producer/mod.rs]: is there any particular reason to have these two `mod`s at the end of the file? i'd prefer them somewhere at the top.

**Comment by xitep** [comment]
@dead10ck this looks great now! 👍 many thanks for all the changes requested. i had just some two stylistic remarks (nothing critical.) feel free to squash the commits as you mentioned and merge it to master.

it's really great to have the ability to write tests against a really kafka broker. i hope we'll grow the suite over time.

btw: i noticed that it takes travis quite a while to finish. it's not critical, but do you think there is anything we could do to speed it up? (maybe running on rust beta is not really beneficial; i certainly do not use it.)

File: src/client/state.rs (hunk 3 above)
Comment (dead10ck) [inline src/client/state.rs]: Hmm, this is a little strange. It does still do this for me, though. Maybe you have an older rustfmt? Anyway, it is a little weird, but harmless to style, I think.

File: tests/integration/consumer_producer/mod.rs (hunk 4 above)
Comment (dead10ck) [inline tests/integration/consumer_producer/mod.rs]: Sure, I'll move them to the top.

**Comment by dead10ck** [comment]
Ok, I squashed, and I'll go ahead and merge. Yeah, it does take a couple of minutes, I think mostly because Travis doesn't keep machine state around—currently, it has to install Rust, then travis-cargo, then download all the Docker images, bring them up and wait for topics to be created, etc. All the different matrices run in parallel, so I don't think removing beta will help much. Not sure if there's much we can do, but we can come back to it later if it becomes a problem.

File: src/client/state.rs (hunk 3 above)
Comment (xitep) [inline src/client/state.rs]: indeed, it was an outdated rustfmt installation.

**Pull request merged**